sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/reflect/DoFnSignaturesTest.java

```diff
@@ -120,4 +130,86 @@ public void processElement() {}
           void finishBundle() {}
         }.getClass());
   }
+
+  @Test
+  public void testStateIdWithWrongType() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("StateId");
+    thrown.expectMessage("StateSpec");
+    DoFnSignatures.INSTANCE.getOrParseSignature(
+        new DoFn<String, String>() {
+          @StateId("foo")
+          String bizzle = "bazzle";
+
+          @ProcessElement
+          public void foo(ProcessContext context) {}
+        }.getClass());
+  }
+
+  @Test
+  public void testStateIdDuplicate() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("Duplicate");
+    thrown.expectMessage("StateId");
+    thrown.expectMessage("my-state-id");
+    thrown.expectMessage("myfield1");
+    thrown.expectMessage("myfield2");
+    DoFnSignature sig =
+        DoFnSignatures.INSTANCE.getOrParseSignature(
+            new DoFn<KV<String, Integer>, Long>() {
+              @StateId("my-state-id")
+              StateSpec<Object, ValueState<Integer>> myfield1 = StateSpecs.value(VarIntCoder.of());
+
+              @StateId("my-state-id")
+              StateSpec<Object, ValueState<Long>> myfield2 = StateSpecs.value(VarLongCoder.of());
+
+              @ProcessElement
+              public void foo(ProcessContext context) {}
+            }.getClass());
+  }
+
+  @Test
+  public void testSimpleStateIdAnonymousDoFn() throws Exception {
+    DoFnSignature sig = DoFnSignatures.INSTANCE.getOrParseSignature(
+        new DoFn<KV<String, Integer>, Long>() {
+          @StateId("foo")
+          StateSpec<Object, ValueState<Integer>> bizzle = StateSpecs.value(VarIntCoder.of());
+
+          @ProcessElement
+          public void foo(ProcessContext context) {}
+        }.getClass());
+
+    assertThat(sig.stateDeclarations().size(), equalTo(1));
+    DoFnSignature.StateDeclaration decl = sig.stateDeclarations().get("foo");
+
+    assertThat(decl.id(), equalTo("foo"));
+    assertThat(decl.field().getName(), equalTo("bizzle"));
+    assertThat(
+        decl.stateType(),
+        Matchers.<TypeDescriptor<?>>equalTo(new TypeDescriptor<ValueState<Integer>>() {}));
+  }
+
+  @Test
+  public void testSimpleStateIdNamedDoFn() throws Exception {
+    DoFnSignature sig =
+        DoFnSignatures.INSTANCE.getOrParseSignature(DoFnForTestSimpleStateIdNamedDoFn.class);
```

Comment:
Just below. Added a comment as a pointer. I could interleave classes with tests, though I end up finding that less readable most of the time.